File: research/experiments/sweep/runs/opt-small-kotlin-sonnet45/src/test/kotlin/optimization/ResultTypesTest.kt
package optimization

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue
import kotlin.test.assertFalse
import kotlin.test.assertNull

class ResultTypesTest {
    @Test
    fun testDefaultOptions() {
        val defaults = defaultOptions()
        assertEquals(1e-8, defaults.gradTol)
        assertEquals(1e-8, defaults.stepTol)
        assertEquals(1e-12, defaults.funcTol)
        assertEquals(1000, defaults.maxIterations)
    }

    @Test
    fun testDefaultOptionsWithOverrides() {
        val opts = defaultOptions { copy(gradTol = 1e-4) }
        assertEquals(1e-4, opts.gradTol)
        assertEquals(1e-8, opts.stepTol)
        assertEquals(1e-12, opts.funcTol)
        assertEquals(1000, opts.maxIterations)
    }

    @Test
    fun testCheckConvergenceGradient() {
        val defaults = defaultOptions()
        val reason = checkConvergence(1e-9, 0.1, 0.1, 5, defaults)
        assertTrue(reason is ConvergenceReason.Gradient)
    }

    @Test
    fun testCheckConvergenceStep() {
        val defaults = defaultOptions()
        val reason = checkConvergence(0.1, 1e-9, 0.1, 5, defaults)
        assertTrue(reason is ConvergenceReason.Step)
    }

    @Test
    fun testCheckConvergenceFunction() {
        val defaults = defaultOptions()
        val reason = checkConvergence(0.1, 0.1, 1e-13, 5, defaults)
        assertTrue(reason is ConvergenceReason.Function)
    }

    @Test
    fun testCheckConvergenceMaxIterations() {
        val defaults = defaultOptions()
        val reason = checkConvergence(0.1, 0.1, 0.1, 1000, defaults)
        assertTrue(reason is ConvergenceReason.MaxIterations)
    }

    @Test
    fun testCheckConvergenceNone() {
        val defaults = defaultOptions()
        val reason = checkConvergence(0.1, 0.1, 0.1, 5, defaults)
        assertNull(reason)
    }

    @Test
    fun testIsConvergedGradient() {
        assertTrue(isConverged(ConvergenceReason.Gradient))
    }

    @Test
    fun testIsConvergedStep() {
        assertTrue(isConverged(ConvergenceReason.Step))
    }

    @Test
    fun testIsConvergedFunction() {
        assertTrue(isConverged(ConvergenceReason.Function))
    }

    @Test
    fun testIsConvergedMaxIterations() {
        assertFalse(isConverged(ConvergenceReason.MaxIterations))
    }

    @Test
    fun testIsConvergedLineSearchFailed() {
        assertFalse(isConverged(ConvergenceReason.LineSearchFailed))
    }

    @Test
    fun testConvergenceMessages() {
        assertEquals("Gradient norm below tolerance", convergenceMessage(ConvergenceReason.Gradient))
        assertEquals("Step size below tolerance", convergenceMessage(ConvergenceReason.Step))
        assertEquals("Function change below tolerance", convergenceMessage(ConvergenceReason.Function))
        assertEquals("Maximum iterations reached", convergenceMessage(ConvergenceReason.MaxIterations))
        assertEquals("Line search failed", convergenceMessage(ConvergenceReason.LineSearchFailed))
    }

    @Test
    fun testConvergencePriority() {
        // When multiple criteria are met, should return first match: gradient → step → function
        val defaults = defaultOptions()

        // All three main criteria met: should return gradient
        val reason1 = checkConvergence(1e-9, 1e-9, 1e-13, 5, defaults)
        assertTrue(reason1 is ConvergenceReason.Gradient)

        // Step and function met: should return step
        val reason2 = checkConvergence(0.1, 1e-9, 1e-13, 5, defaults)
        assertTrue(reason2 is ConvergenceReason.Step)
    }
}

File: research/experiments/sweep/runs/opt-small-typescript-sonnet45/src/result-types.test.ts
import { test, expect, describe } from "bun:test";
import {
  defaultOptions,
  checkConvergence,
  isConverged,
  convergenceMessage,
  type ConvergenceReason,
} from "./result-types";

describe("result-types", () => {
  describe("defaultOptions", () => {
    test("returns default values", () => {
      const opts = defaultOptions();
      expect(opts.gradTol).toBe(1e-8);
      expect(opts.stepTol).toBe(1e-8);
      expect(opts.funcTol).toBe(1e-12);
      expect(opts.maxIterations).toBe(1000);
    });

    test("applies overrides", () => {
      const opts = defaultOptions({ gradTol: 1e-4 });
      expect(opts.gradTol).toBe(1e-4);
      expect(opts.stepTol).toBe(1e-8);
      expect(opts.funcTol).toBe(1e-12);
      expect(opts.maxIterations).toBe(1000);
    });
  });

  describe("checkConvergence", () => {
    const defaults = defaultOptions();

    test("returns gradient when gradient norm below tolerance", () => {
      const reason = checkConvergence(1e-9, 0.1, 0.1, 5, defaults);
      expect(reason).toEqual({ kind: "gradient" });
    });

    test("returns step when step norm below tolerance", () => {
      const reason = checkConvergence(0.1, 1e-9, 0.1, 5, defaults);
      expect(reason).toEqual({ kind: "step" });
    });

    test("returns function when function change below tolerance", () => {
      const reason = checkConvergence(0.1, 0.1, 1e-13, 5, defaults);
      expect(reason).toEqual({ kind: "function" });
    });

    test("returns maxIterations when iteration limit reached", () => {
      const reason = checkConvergence(0.1, 0.1, 0.1, 1000, defaults);
      expect(reason).toEqual({ kind: "maxIterations" });
    });

    test("returns null when no criterion met", () => {
      const reason = checkConvergence(0.1, 0.1, 0.1, 5, defaults);
      expect(reason).toBe(null);
    });

    test("prioritizes gradient over other criteria", () => {
      // All criteria met, should return gradient first
      const reason = checkConvergence(1e-9, 1e-9, 1e-13, 1000, defaults);
      expect(reason).toEqual({ kind: "gradient" });
    });

    test("prioritizes step over function and maxIterations", () => {
      // Step, function, and maxIterations all met
      const reason = checkConvergence(0.1, 1e-9, 1e-13, 1000, defaults);
      expect(reason).toEqual({ kind: "step" });
    });

    test("prioritizes function over maxIterations", () => {
      // Function and maxIterations both met
      const reason = checkConvergence(0.1, 0.1, 1e-13, 1000, defaults);
      expect(reason).toEqual({ kind: "function" });
    });
  });

  describe("isConverged", () => {
    test("returns true for gradient", () => {
      expect(isConverged({ kind: "gradient" })).toBe(true);
    });

    test("returns true for step", () => {
      expect(isConverged({ kind: "step" })).toBe(true);
    });

    test("returns true for function", () => {
      expect(isConverged({ kind: "function" })).toBe(true);
    });

    test("returns false for maxIterations", () => {
      expect(isConverged({ kind: "maxIterations" })).toBe(false);
    });

    test("returns false for lineSearchFailed", () => {
      expect(isConverged({ kind: "lineSearchFailed" })).toBe(false);
    });
  });

  describe("convergenceMessage", () => {
    test("returns message for each kind", () => {
      expect(convergenceMessage({ kind: "gradient" })).toContain("gradient");
      expect(convergenceMessage({ kind: "step" })).toContain("step");
      expect(convergenceMessage({ kind: "function" })).toContain("function");
      expect(convergenceMessage({ kind: "maxIterations" })).toContain(
        "Maximum iterations"
      );
      expect(convergenceMessage({ kind: "lineSearchFailed" })).toContain(
        "Line search"
      );
    });
  });
});
